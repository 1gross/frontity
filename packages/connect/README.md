## Description
The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.

You can find more information about it at [Frontity docs](https://docs.frontity.org/javascript-and-and-react/react-basic#state-manager-frontity-connect).
