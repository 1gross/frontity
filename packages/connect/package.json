{
  "name": "@frontity/connect",
  "version": "0.1.0",
  "description": "A state manager that's easy to use and hackable at the same time.",
  "keywords": [
    "frontity",
    "connect"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "jest",
    "test": "jest --watch",
    "test:inspect": "node --inspect ./node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "recursive-iterator": "^3.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@types/jest": "^24.0.11",
    "babel-jest": "^24.8.0",
    "jest": "^24.7.1",
    "ts-jest": "^24.0.2",
    "typescript": "^3.4.2"
  }
}
