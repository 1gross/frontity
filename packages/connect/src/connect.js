import { createContext, useContext, createElement } from "react";
import { view } from "@risingstack/react-easy-state";

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext();

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * Connects a React component with the Frontity connect store.
 *
 * @param Component The React component to be connected
 * @returns The same Component, but with the store included in the props.
 */
export const connect = (Component) => {
  const isStatelessComp = !(
    Component.prototype && Component.prototype.isReactComponent
  );

  if (isStatelessComp) {
    return view((props) => {
      const store = useContext(context);
      return Component({ ...props, ...store });
    });
  } else {
    class ConnectedComponent extends Component {
      static contextType = context;

      render() {
        return createElement(Component, {
          ...this.props,
          ...this.context,
        });
      }
    }

    return view(ConnectedComponent);
  }
};
