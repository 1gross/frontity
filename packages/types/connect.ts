import Package from "./package";
import { ResolveState, Omit } from "./utils";

// Resolves the actions for things like Action and Derived.
export type ResolveActions<Actions extends Package["state"]> = {
  [P in keyof Actions]: Actions[P] extends (
    ...a: any
  ) => (arg: infer Arg) => void // Turns "state => args => {}" into "args => {}"
    ? (arg: Arg) => void // Turns "state => {}" into "() => {}"
    : Actions[P] extends (state: Package["state"]) => any
    ? () => void
    : ResolveActions<Actions[P]>
};

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export default Connect;
