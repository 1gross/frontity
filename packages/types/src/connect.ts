import Package from "./package";
import { ResolveState, Omit, ResolveActions } from "./utils";

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export type ExternalProps<T extends Package> = Omit<
  T,
  "state" | "actions" | "libraries" | "roots" | "fills"
>;

export default Connect;
