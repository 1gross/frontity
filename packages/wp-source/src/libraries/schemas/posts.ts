import { schema } from "normalizr";
import { taxonomies } from "./taxonomies";
import { author } from "./authors";
import { attachment } from "./attachments";
import { normalize } from "../route-utils";

export const postType = new schema.Entity("postType");

export const post = new schema.Entity(
  "post",
  {},
  {
    processStrategy(entity) {
      const result = { ...entity };
      result.link = normalize(result.link);
      return result;
    }
  }
);

post.define({
  _embedded: {
    author: [author],
    type: [postType],
    "wp:featuredmedia": [attachment],
    "wp:contentmedia": [[attachment]],
    "wp:term": taxonomies
  }
});

export const posts = new schema.Array(post);
