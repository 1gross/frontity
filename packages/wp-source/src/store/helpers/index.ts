export const getIdBySlug = (
  entityMap: {
    [id: number]: {
      slug: string;
      id: number;
    };
  },
  slug: string
): number | undefined => {
  const entity = Object.values(entityMap).find(e => e.slug === slug);
  return entity && entity.id;
};

export const getTotal = (response: Response): number =>
  parseInt(response.headers.get("X-WP-Total"));

export const getTotalPages = (response: Response): number =>
  parseInt(response.headers.get("X-WP-TotalPages"));

export const normalizePath = (pathOrLink: string): string => {
  try {
    return new URL(pathOrLink).pathname;
  } catch (e) {
    // is not a URL
    return pathOrLink;
  }
};
