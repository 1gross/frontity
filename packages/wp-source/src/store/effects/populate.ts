import { normalize } from "normalizr";
import * as schemas from "../../schemas";
import { Context, DataPage } from "../../types";

export const populate = async (
  state: Context["state"]["source"],
  response: Response
): Promise<DataPage> => {
  // Normalize response
  const json = await response.json();
  const isList = json instanceof Array;
  const { entities, result } = normalize(
    json,
    isList ? schemas.list : schemas.entity
  );

  // type, id and link of added entities
  const entityList: DataPage = (isList ? result : [result]).map(
    ({ id: entityId, schema }) => {
      const { type, id, link } = entities[schema][entityId];
      return { type, id, link };
    }
  );

  // add entities to state
  Object.entries(entities).forEach(([schema, entityMap]) => {
    Object.entries(entityMap).forEach(([, entity]) => {
      if (schema === "postType" || schema === "attachment") {
        if (state[entity.type]) state[entity.type][entity.id] = entity;
      } else if (schema === "taxonomy") {
        if (state[entity.taxonomy]) state[entity.taxonomy][entity.id] = entity;
      } else if (schema === "author") {
        state.author[entity.id] = entity;
      }
    });
  });

  return entityList;
};
