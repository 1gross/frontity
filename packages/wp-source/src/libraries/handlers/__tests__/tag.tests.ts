import { createStore, InitializedStore } from "@frontity/connect";
import WpSource from "../../../../types";
import populate from "../../populate";
import handler from "../tag";
import { mockResponse, expectEntities } from "./mocks/helpers";
import posts from "./mocks/posts-tag-9.json";

import wpSource from "../../../";
jest.mock("../../../");

let store: InitializedStore<WpSource>;

beforeEach(() => {
  // First, get a mocked instance of wpSource
  const config = wpSource();

  // Then, replaces the mocked implementation of Api
  const { api } = config.libraries.source;
  api.getIdBySlug = jest.fn().mockResolvedValue(9);
  api.get = jest.fn().mockResolvedValue(
    mockResponse(posts, {
      "X-WP-Total": 3,
      "X-WP-TotalPages": 1
    })
  );

  // And add the original implementation of populate
  config.libraries.source.populate = jest.fn().mockImplementation(populate);

  // Instantiate the store
  store = createStore(config);
});

describe("tag", () => {
  test("doesn't exist in source.tag", async () => {
    // source.fetch("/tag/iceland/")
    store.state.source.data["/tag/iceland/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/tag/iceland/",
      params: { slug: "iceland" },
      ...store
    });

    expect(store.state.source.data).toMatchSnapshot();
    expectEntities(store.state.source);
  });

  test("exists in source.tag but not in source.data", async () => {
    store.state.source.tag[9] = {
      id: 9,
      count: 3,
      description: "",
      link: "https://test.frontity.io/tag/iceland/",
      name: "Iceland",
      slug: "iceland",
      taxonomy: "tag",
      meta: []
    };

    // source.fetch("/tag/iceland/")
    store.state.source.data["/tag/iceland/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/tag/iceland/",
      params: { slug: "iceland" },
      ...store
    });

    expect(store.libraries.source.api.getIdBySlug).not.toBeCalled();
    expect(store.state.source.data).toMatchSnapshot();
    expectEntities(store.state.source);
  });

  test("fetchs from a different endpoint with extra params", async () => {
    store.state.source.postEndpoint = "multiple-post-type";
    store.state.source.params = { type: ["post", "travel"] };
    // source.fetch("/tag/iceland/")
    store.state.source.data["/tag/iceland/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/tag/iceland/",
      params: { slug: "iceland" },
      ...store
    });

    const apiGet = jest.spyOn(store.libraries.source.api, "get");
    expect(apiGet.mock.calls).toMatchSnapshot();
  });
});
