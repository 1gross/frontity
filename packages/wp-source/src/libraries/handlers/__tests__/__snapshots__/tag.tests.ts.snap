// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tag doesn't exist in source.tag 1`] = `
Object {
  "/tag/iceland/": Object {
    "id": 9,
    "isArchive": true,
    "isFetching": true,
    "isReady": false,
    "isTag": true,
    "isTaxonomy": true,
    "items": Array [
      Object {
        "id": 60,
        "link": "/2016/the-beauties-of-gullfoss/",
        "type": "post",
      },
      Object {
        "id": 26,
        "link": "/2016/discovering-iceland/",
        "type": "post",
      },
      Object {
        "id": 1,
        "link": "/2016/iceland/",
        "type": "post",
      },
    ],
    "taxonomy": "tag",
    "total": 3,
    "totalPages": 1,
  },
}
`;

exports[`tag doesn't exist in source.tag 2`] = `
Object {
  "attachment": Array [
    6,
    27,
    62,
  ],
  "author": Array [
    3,
    4,
  ],
  "category": Array [
    3,
    4,
    6,
    7,
    8,
    56,
    57,
    58,
    59,
    108,
  ],
  "post": Array [
    1,
    26,
    60,
  ],
  "tag": Array [
    9,
    10,
    11,
    13,
    40,
    47,
  ],
}
`;

exports[`tag exists in source.tag but not in source.data 1`] = `
Object {
  "/tag/iceland/": Object {
    "id": 9,
    "isArchive": true,
    "isFetching": true,
    "isReady": false,
    "isTag": true,
    "isTaxonomy": true,
    "items": Array [
      Object {
        "id": 60,
        "link": "/2016/the-beauties-of-gullfoss/",
        "type": "post",
      },
      Object {
        "id": 26,
        "link": "/2016/discovering-iceland/",
        "type": "post",
      },
      Object {
        "id": 1,
        "link": "/2016/iceland/",
        "type": "post",
      },
    ],
    "taxonomy": "tag",
    "total": 3,
    "totalPages": 1,
  },
}
`;

exports[`tag exists in source.tag but not in source.data 2`] = `
Object {
  "attachment": Array [
    6,
    27,
    62,
  ],
  "author": Array [
    3,
    4,
  ],
  "category": Array [
    3,
    4,
    6,
    7,
    8,
    56,
    57,
    58,
    59,
    108,
  ],
  "post": Array [
    1,
    26,
    60,
  ],
  "tag": Array [
    9,
    10,
    11,
    13,
    40,
    47,
  ],
}
`;

exports[`tag fetchs from a different endpoint with extra params 1`] = `
Array [
  Array [
    Object {
      "endpoint": "multiple-post-type",
      "params": Object {
        "_embed": true,
        "page": 1,
        "search": undefined,
        "tags": 9,
        "type": Array [
          "post",
          "travel",
        ],
      },
    },
  ],
]
`;
